File: typescript/src/lexer.ts
import { Token, createToken, TokenType, TokenTypes } from "./token";

export class Lexer {
    private currPos: number;
    private nextPos: number;
    private char: string;

    private keywords: ReadonlyMap<string, TokenType> = new Map([
        ["fn", TokenTypes.FUNCTION],
        ["let", TokenTypes.LET],
    ]);

    constructor(private input: string) {
        this.currPos = 0;
        this.nextPos = 1;
        this.char = this.input[0];
    }

    nextToken(): Token {
        this.skipWhitespace();

        //Tokenizer
        let token: Token;
        switch (this.char) {
            case '=' :
                token = createToken(TokenTypes.ASSIGN, this.char);
            break;
            case ';':
                token = createToken(TokenTypes.SEMICOLON, this.char);
            break;
            case '(':
                   token = createToken(TokenTypes.LPAREN, this.char);
            break;
            case ')':
                   token = createToken(TokenTypes.RPAREN, this.char);
            break;
            case ',':
                token = createToken(TokenTypes.COMMA, this.char);
            break;
            case '+':
                token = createToken(TokenTypes.PLUS, this.char);
            break;
            case '{':
                token = createToken(TokenTypes.LBRACE, this.char);
            break;
            case '}':
                token = createToken(TokenTypes.RBRACE, this.char);
            break;
            case '\0':
                token = createToken(TokenTypes.EOF, '');
            break;
            default:
                if(isLetter(this.char)) {
                    const ident = this.readIdentifier();
                    token = createToken(this.lookupKeywords(ident), ident);

                    return token;
                }

                if(isNumber(this.char)) {
                    const numb = this.readNumber();
                    token = createToken(TokenTypes.INT, numb);

                    return token;
                }
                token = createToken(TokenTypes.ILLEGAL, this.char);
        };

        this.readChar();

        return token;
    }

    private readChar() {
        if(this.nextPos >= this.input.length) {
            this.char = '\0';
        } else {
            this.char = this.input[this.nextPos];
        }
        this.currPos = this.nextPos;
        this.nextPos++;
    }

    private iterate(condition: (char: string) => boolean) {
        const startPos = this.currPos;

        while(condition(this.char)) {
            this.readChar();
        }

        return this.input.substring(startPos, this.currPos);
    }

    private readIdentifier(): string {
        return this.iterate(isLetter);
    }

    //NOTE: Ignoring float numbers
    private readNumber(): string {
        return this.iterate(isNumber);
    }

    private lookupKeywords(identifier: string): TokenType {
          const keyword = this.keywords.get(identifier);
          return keyword ? keyword : TokenTypes.IDENT;
    }

    private skipWhitespace(): void {
        if(this.char === ' ' || this.char === '\t' || this.char === '\r' || this.char === '\n') {
            this.readChar();
            this.skipWhitespace();
        }
    }
}

function isLetter(char: string): boolean {
    return char >= 'a' && char <= 'z' || char >= 'A' && char <= 'Z' || char === '_';
}

function isNumber(char: string): boolean {
    return char >= '0' && char <= '9';
}

File: java/src/main/java/interpreter/Lexer.java
package interpreter;

import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

public class Lexer {
    private String input;
    private int position;
    private int nextPos;
    private char ch;
    private static final Map<String, Token> keywords;
    static {
        Map<String, Token> map = new HashMap<String, Token>();
        map.put(Token.LET.getLiteral(), Token.LET);
        map.put(Token.FUNCTION.getLiteral(), Token.FUNCTION);
        map.put(Token.IF.getLiteral(), Token.IF);
        map.put(Token.ELSE.getLiteral(), Token.ELSE);
        map.put(Token.TRUE.getLiteral(), Token.TRUE);
        map.put(Token.FALSE.getLiteral(), Token.FALSE);
        map.put(Token.RETURN.getLiteral(), Token.RETURN);
        keywords = Collections.unmodifiableMap(map);
    }

    public Lexer(String input) {
        this.input = input;
        this.position = 0;
        this.nextPos = getNextPos();
        this.ch = getChar();
    }

    public Token nextToken() {
        skipWhiteSpace();

        Token tok;

        switch (this.ch) {
            case '=':
                if (peek() == '=') {
                    tok = Token.EQ;
                    readChar();
                    readChar();
                    return tok;
                }
                tok = Token.ASSIGN;
                break;
            case '!':
                if (peek() == '=') {
                    tok = Token.NOT_EQ;
                    readChar();
                    readChar();
                    return tok;
                }
                tok = Token.BANG;
                break;
            case ';':
                tok = Token.SEMICOLON;
                break;
            case '(':
                tok = Token.LPAREN;
                break;
            case ')':
                tok = Token.RPAREN;
                break;
            case ',':
                tok = Token.COMMA;
                break;
            case '+':
                tok = Token.PLUS;
                break;
            case '-':
                tok = Token.MINUS;
                break;
            case '/':
                tok = Token.SLASH;
                break;
            case '<':
                tok = Token.LT;
                break;
            case '>':
                tok = Token.GT;
                break;
            case '*':
                tok = Token.ASTERISK;
                break;
            case '{':
                tok = Token.LBRACE;
                break;
            case '}':
                tok = Token.RBRACE;
                break;
            case '\0':
                tok = Token.EOF;
                break;
            default:
                if (TypeChecker.isLetter(this.ch)) {
                    String ident = readIdentifier();
                    Token keywordToken = keywords.get(ident);

                    if (keywordToken == null) {
                        tok = Token.IDENT.literal(ident);
                    } else {
                        tok = keywordToken;
                    }
                    return tok;
                }

                if (TypeChecker.isDigit(this.ch)) {
                    String number = readNumber();
                    return Token.INT.literal(number);
                }

                tok = Token.ILLEGAL;
        }

        this.readChar();
        return tok;
    }

    private char getChar() {
        return input.charAt(position);
    }

    private int getNextPos() {
        return position + 1;
    }

    private char peek() {
        return input.charAt(nextPos);
    }

    private void readChar() {
        this.position = nextPos;
        this.nextPos = getNextPos();

        if (position >= input.length()) {
            this.ch = '\0';
        } else {
            this.ch = getChar();
        }
    }

    private String readIdentifier() {
        int start = this.position;

        while (TypeChecker.isLetter(this.ch)) {
            readChar();
        }

        return input.substring(start, this.position);
    }

    private String readNumber() {
        int start = this.position;

        while (TypeChecker.isDigit(this.ch)) {
            readChar();
        }

        return input.substring(start, this.position);
    }

    private void skipWhiteSpace() {
        if (this.ch == ' ' || this.ch == '\t' || this.ch == '\r' || this.ch == '\n') {
            readChar();
            skipWhiteSpace();
        }
    }
}
